### PR TITLE
Introduces the TimeLayerSliderPanel for WMS-T layer

### PR DESCRIPTION
**Pull request by mholthausen**
# FEATURE

## Description

This merges the existing `TimeSlider` and `TimeLayerAware` components into the new panel component `TimeLayerSliderPanel`.

timelayersliderpanel

Some points are still open:

* [x] [tooltips](https://github.com/terrestris/react-geo/compare/master...mholthausen:introduce-timelayersliderpanel?expand=1#diff-0e03078ed2beb0b1d78fb55d39828718R476): object access via string literals is disallowed
* [x] [ts-ignore](https://github.com/terrestris/react-geo/compare/master...mholthausen:introduce-timelayersliderpanel?expand=1#diff-0e03078ed2beb0b1d78fb55d39828718R183): When removed, the application fails to compile `Argument of type 'typeof TimeSlider' is not assignable to parameter of type 'ComponentType<TimeSliderProps>'.`
* [x] Tests: no working tests included until yet

Maybe another rect-geo @terrestris/devs can help me out here soon.

### CODE REVIEW

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+
+export interface TimeLayerSliderPanelProps extends Partial<DefaultTimeLayerSliderPanelProps> {
+  map: any;
+  initStartDate: moment.Moment;
+  initEndDate: moment.Moment;
+}
+
+export interface TimeLayerSliderPanelState {
+  value: moment.Moment;
+  playbackSpeed: string;
```
Comment (dnlkoch) [inline src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: I would expect a `number` here.

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+  startDate: moment.Moment;
+  endDate: moment.Moment;
+}
+
+/**
+ * mapStateToProps - mapping state to props of Map Component.
+ *
+ * @param {Object} state current state
+ * @return {Object} mapped props
+ */
```
Comment (dnlkoch) [inline src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: Please remove unneeded multiline comment.

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+  /**
+   * The default props of LayerSetBaseMapChooser
+   *
+   * @static
+   * @type {DefaultLayerSetBaseMapChooserProps}
+   * @memberof LayerSetBaseMapChooser
+   */
+  public static defaultProps: DefaultTimeLayerSliderPanelProps = {
+    className: '',
+    onChange: () => {/* empty */},
```
Comment (dnlkoch) [inline src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]:
```suggestion
    onChange: () => {},
```

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+
+    this.setState({
+      startDate: initStartDate,
+      endDate: initEndDate
+    });
+  }
+
+  componentDidUpdate(prevProps: TimeLayerSliderPanelProps) {
+    // TODO: this deep check may impact performance..
+    if (!(_isEqual(prevProps.timeAwareLayers, this.props.timeAwareLayers))) {
```
Comment (dnlkoch) [inline src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: You may want to compare the `ol_uid` of each layer instead.

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+      // update slider properties if some another layer set was chosen
```
Comment (dnlkoch) [inline src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]:
```suggestion
      // update slider properties if layers were updated
```

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+    if (nextState.startDate !== startDate) {
+      return true;
+    }
+    if (nextState.endDate !== endDate) {
+      return true;
+    }
+    if (!(_isEqual(nextProps.timeAwareLayers, timeAwareLayers))) {
+      return true;
+    }
+    return false;
```
Comment (dnlkoch) [inline src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: I like the approach, but please also add update support for at least the `tooltips` prop.

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx [hunk 7]
```diff
@@ -0,0 +1,496 @@
...
+    if (this.props.onChange) {
+      this.props.onChange(newValue);
+    }
+  }
+
+  /**
+   * makes sure that the appended time parameter in GetMap calls
+   * is rounded to full hours to receive a valid response
+   */
+  wmsTimeHandler = (value?: any) => {
```
Comment (dnlkoch) [inline src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: Why do we need this? A WMS-Time may also accept a non-full-hours time parameter, see e.g. [here](https://docs.geoserver.org/stable/en/user/services/wms/time.html).

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+        transform: 'translate(50%)'
+      }
+    };
+    marks[mid.toISOString()] = {
+      label: mid.format(dateFormat)
+    };
+
+    return (
+      <div className={`time-layer-slider ${disabledCls}`.trim()}>
+
```
Comment (dnlkoch) [inline src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: Please remove empty line.

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx [hunk 9]
```diff
@@ -0,0 +1,496 @@
...
+          className={extraCls + ' speed-picker'}
+          onChange={this.onPlaybackSpeedChange}
+        >
+          <Option value="0.5">0.5x</Option>
+          <Option value="1">1x</Option>
+          <Option value="2">2x</Option>
+          <Option value="5">5x</Option>
+          <Option value="10">10x</Option>
+          <Option value="100">100x</Option>
+          <Option value="300">300x</Option>
```
Comment (dnlkoch) [inline src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: It would be nice to have this list configurable.

File: src/Slider/TimeSlider/TimeSlider.tsx [hunk 10]
```diff
@@ -52,7 +52,7 @@ export interface BaseProps {
   marks?: SliderMarks;
 }
 
-export type TimeSliderProps = BaseProps & Partial<DefaultProps> & SliderProps;
```
Comment (dnlkoch) [inline src/Slider/TimeSlider/TimeSlider.tsx]: Having the `TimeSliderProps` extend the `SliderProps` seems very legit to me, why did you remove them?

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx (hunk 2 above)
Comment (KaiVolland) [reply src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: … also remove `{Object}`. The types are determined by TypeScript now.

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+
+  private _TimeLayerAwareSlider: any;
+  private _wmsTimeLayers: any[];
+  private _interval: number;
+
+  /**
+   * The default props of LayerSetBaseMapChooser
+   *
+   * @static
+   * @type {DefaultLayerSetBaseMapChooserProps}
```
Comment (KaiVolland) [inline src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]:
```suggestion
```

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx [hunk 12]
```diff
@@ -0,0 +1,496 @@
...
+    });
+    newStartDate = startDatesFromLayers.length > 0 ? moment.min(startDatesFromLayers) : startDate;
+    newEndDate = endDatesFromLayers.length > 0 ? moment.max(endDatesFromLayers) : endDate;
+    this.updateDataRange([newStartDate, newEndDate]);
+  }
+
+  /**
+   * Handler for the time slider change behaviour
+   */
+  timeSliderCustomHandler = (value: any) => {
```
Comment (KaiVolland) [inline src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: The type of the SliderHandler should be known. Maybe an Event or number array.

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx (hunk 3 above)
Comment (KaiVolland) [reply src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: I think it should be `onChange: (value: moment.Moment) => void,`

Comment (KaiVolland) [reply src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: If it is empty it shouldn't be a `defaultProp`

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx (hunk 1 above)
Comment (mholthausen) [reply src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: Th playback speed can also be set to `moment()` values like `'hours'`, `'days'`, `'months'` etc. Since these are then passed as `string`, the entire data type of this variable has probably been set to the type of `string`.

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx (hunk 3 above)
Comment (mholthausen) [reply src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: If it is set like suggested, the linting fails:` no-empty: block is empty`. The proposed solutions was to enter an comment into this block.
If these are removed completely, compilation fails, because this information is missing in `TimeLayerAware`. Thus, I would suggest to leave them as they are, or do you have another suggestion @KaiVolland ?

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx (hunk 4 above)
Comment (mholthausen) [reply src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: Comparing the `ol_uid` of each layer in `timeAwareLayers` of `prevProps` against the layer at the same position of `this.props`

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx (hunk 7 above)
Comment (mholthausen) [reply src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: I don't know, it was implemented that way in the react-geo-baseclient. It may have been needed to display images for a timespan of a complete hour, if there is only one image available for a solid hour. I have now added a query to the configuration of the layer, whether this behavior is desired or not (`roundToFullHours`).

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx (hunk 9 above)
Comment (mholthausen) [reply src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: `autoPlaySpeedOptions` can be passed to the panel now. If these are not specified, default values are used.

File: src/Slider/TimeSlider/TimeSlider.tsx (hunk 10 above)
Comment (mholthausen) [reply src/Slider/TimeSlider/TimeSlider.tsx]: We have removed these, because otherwise an error occurs with `timeLayerAware(TimeSlider, this._wmsTimeLayers);`: `Argument of type 'typeof TimeSlider' is not assignable to parameter of type 'ComponentType<TimeSliderProps>'`. This could only be bypassed for compilation with the specification `\\ @ts-ignore`

File: src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx (hunk 3 above)
Comment (mholthausen) [reply src/Panel/TimeLayerSliderPanel/TimeLayerSliderPanel.tsx]: Set the `tslint.json` to allow empty functions.

File: src/Slider/TimeSlider/TimeSlider.tsx (hunk 10 above)
Comment (mholthausen) [reply src/Slider/TimeSlider/TimeSlider.tsx]: The doubled props defined by the antd slider are omitted.
